src/regressors.py

```diff
@@ -0,0 +1,428 @@
+
+
+
+
+class RegressionModel():
+    """
+    This is an abstract class to describe regression models.
+    They have 3 core methods.
+    1. Initialisation with correct hyperparameters.
+    2. Fitting with training data.
+    3. Predicting values.
+    They all predict 1D outputs only.
+    Some models can take in a variety of different input dimensions.
+    """
+    def __init__(self, num_input_dim :int, min_length :int) -> None:
+        """
+        All Regression models use a scaler to rescale the input data.
+        And have a regressor.
+        The number of dimensions is specified to mitigate errors.
+
+        Args:
+            num_input_dim (int): number of dimensions/features of training (and test) data.
+            min_length (int): minimum length of training dataset.
+        """
+        self._scaler = preprocessing.StandardScaler()
+        self._regressor = None
+        self._x_dim = num_input_dim
+        self._min_length = min_length
+
+
+    def prepare_fit(self, train_x :np.ndarray[float], train_y :np.ndarray[float]) -> np.ndarray[float]:
+        """
+        Checks that the training data is correctly dimensioned and normalises it.
+
+        Args:
+            train_x (np.ndarray[float]): n by d array of n input data values of dimension d.
+            train_y (np.ndarray[float]): n by 1 array of n output data values.
+
+        Returns:
+            np.ndarray[float]: scaled n by d array of n input data values of dimension d.
+        """
+        self.check_dim(len(train_x[0]), self._x_dim, 'Input')
+        self.check_length(len(train_x))
+        if type(train_y[0]) != np.ndarray:
+            raise Exception('Output data should be a numpy array of shape (-1, 1).')
+        self.check_dim(len(train_y[0]), 1, 'Output')
+
+        self._scaler.fit(train_x)
+        return self._scaler.transform(train_x)
+
+
+    def prepare_predict(self, predict_x :np.ndarray[float]) -> np.ndarray[float]:
+        """
+        Checks that the prediction data is correctly dimensioned and normalises it.
+
+        Args:
+            predict_x (np.ndarray[float]): n by d array of n input data values of dimension d.
+
+        Returns:
+            np.ndarray[float]: scaled n by d array of n input data values of dimension d.
+        """
+        self.check_dim(len(predict_x[0]), self._x_dim, 'Input')
+        return self._scaler.transform(predict_x)
+
+
+    def check_dim(self, dim :int, correct_dim :int, data_name :str) -> None:
+        """
+        Checks the dimensions/features are equal to a specified number.
+
+        Args:
+            dim (int): measured number of dimensions.
+            correct_dim (int): expected number of dimensions.
+            data_name (str): name for exception handling.
+
+        Raises:
+            Exception: error to explain user's mistake.
+        """
+        if dim != correct_dim:
+            raise Exception(data_name+' data should be a numpy array of shape (-1, '+str(correct_dim)+').')
```

Comment:
Suggest using f-strings for strings containing parameter values like this, as they are slightly more readable and nicer to write, but ultimately either approach produces the same result.
```suggestion
            raise Exception(f"{data_name} data should be a numpy array of shape (-1, {str(correct_dim)}).")
```

Reply:
Same for several similar instances below.